Task 2 - include unit tests

Bee0933 opened the pull request:
added unit test  & modified class method  

Class modification:
- changed dataset to ny taxi data
- added bash script to download taxi data based on year and date positional arguments 
- added class method to run bash script 
- added class method to preprocess data

Unit test 
- added test for bash script input arguments 
- added mock test for bash script download 
- added test for class instance & mock test for gspread client connection

Commit:
initial commit

Commit:
added output link

Commit:
added error handling using 'raise'

Commit:
removed redudndant error logger

Commit:
added unit test  & modified class method 

- change dataset to ny taxi data
- added bash script to download taxi data with year and date arguments
- added class method to run bash script
- added class method to preprocess data
- added unit tests

elseagle commented on project-1/main.py:
```diff
@@ -0,0 +1,211 @@
+        try:
+            # google cloud platform API credentials
+            creds = service_account.Credentials.from_service_account_file(
+                "./credentials.json", scopes=scope
```
nit: avoid magic strings, imagine we deploy this and this file isn't at "./credentials.json". Use a constant instead

Bee0933 replied:
ohh yeah.. thats true

I'd do that

elseagle commented on project-1/main.py:
```diff
@@ -0,0 +1,211 @@
+        """
+
+        # bash script name
+        bash_script = "extract.sh"
```
nit: let's move this to a constant

elseagle commented on project-1/main.py:
```diff
@@ -0,0 +1,211 @@
+
+
+# Error handlers
+class DataPopulationError(Exception):
```
You can move your exception classes to exceptions.py file

elseagle commented on project-1/main.py:
```diff
@@ -0,0 +1,211 @@
+    wb.create_new_sheet()
+    wb.process_data()
+    wb.populate_sheet_from_csv()
+    # wb.analytics()
```
If this isn't needed, you can remove it

elseagle commented on project-1/tests/unit/test_gspread_utility.py:
```diff
@@ -0,0 +1,77 @@
+import os
+from unittest.mock import patch, Mock
+
+EXTRACT_SH_PATH = "/home/bee_nyah/Desktop/DATA_EPIC/Best_Nyah/project-1/extract.sh"
```
Move to environment variable

elseagle commented on project-1/tests/unit/test_gspread_utility.py:
```diff
@@ -0,0 +1,77 @@
+
+EXTRACT_SH_PATH = "/home/bee_nyah/Desktop/DATA_EPIC/Best_Nyah/project-1/extract.sh"
+
+def subprocess_run(argument: List[str]) -> str:
```
since this isn't a test case, kindly move to an utility or helper file, you can create a folder and script for that in .../test/

elseagle commented on project-1/tests/unit/test_gspread_utility.py:
```diff
@@ -0,0 +1,77 @@
+    Tests file downlad from bash script
+    """
+    # test for bad argument
+    with pytest.raises(subprocess.CalledProcessError):
```
nit: each test case should have at least 2 assertions, but I've seen other test cases having more than one, so I can let this slide

elseagle reviewed (changes requested):
Thanks for the submission, looks solid! Just a few comments

Commit:
refactor code to address requested PR change

- removed magic strings
- created exception.py file for error handlers
- removed redundant code
- created .env for environment variables
- created utils package for test helper functions
- included more accessions

elseagle commented on project-1/main.py:
```diff
@@ -0,0 +1,193 @@
+from pathlib import Path
+from google.oauth2 import service_account
+from gspread_dataframe import set_with_dataframe
+from exceptions import *
```
I'm anti "import *" - LOL, let's see all we're importing, it makes the code easier to read

Bee0933 replied:
sorry for my habit.. LOL

elseagle commented on project-1/main.py:
```diff
@@ -0,0 +1,193 @@
+    handlers=[logging.StreamHandler()],
+)
+
+CREDENTIALS_PATH = "./credentials.json"
```
Nice, we could just move this entirely to the environment as variables, so this makes it very easy to change these values and not bother changing the code in the future, sorry if I didn't make that clear before

Bee0933 replied:
quite true chief ... I'd make changes

elseagle commented on project-1/main.py:
```diff
@@ -0,0 +1,193 @@
+)
+
+CREDENTIALS_PATH = "./credentials.json"
+BASH_SCRIPT_PATH = "./extract.sh"
```
ditto for this

elseagle commented on project-1/main.py:
```diff
@@ -0,0 +1,193 @@
+            stdout=subprocess.PIPE,
+        )
+        script_output = result.stdout
+        download_path = script_output.split("\n")[-2]
```
Can you comment an example for this string extraction, like what is the value we're trying to extract and what does what we finally extract look like

elseagle commented on project-1/tests/unit/__init__.py:
```diff
@@ -0,0 +1 @@
+from . import *
```
Anti-"import *" movement!

elseagle commented on project-1/tests/unit/test_gspread_utility.py:
```diff
@@ -0,0 +1,71 @@
+from ..utils.utils import subprocess_run
+
+
+EXTRACT_SH_PATH = config("EXTRACT_SH_PATH")
```
You're doing this already, why not do it in the main script too?

elseagle reviewed (changes requested):
Awesome! We're almost there, kindly check out the new comments

Commit:
made PR request changes

- avoided anti-import all (*)
- moved path variables to .env
- added code comment

tobaadesugba commented on project-1/main.py:
```diff
@@ -0,0 +1,203 @@
+            raise NewSpreadsheetError("Error creating a new spreadsheet") from e
+
+    # create new work sheet from spreadsheet
+    def create_new_sheet(self) -> None:
```
Don't forget your docstrings

tobaadesugba commented on project-1/main.py:
```diff
@@ -0,0 +1,203 @@
+            raise DataPopulationError("Error populating sheet from CSV") from e
+
+
+def main(
```
You should consider refactoring your function definition like so...

```
def main(spread_sheet: str,
         worksheet_name: str,
         email: str,
         year: str,
         month: str) -> None:
```

tobaadesugba reviewed (changes requested):
Great work, just a few nitpickings.

tobaadesugba commented on project-1/preprocess.ipynb:
```diff
@@ -0,0 +1,524 @@
+{
```
<p>I'm not sure if this preprocessing was used in the data that you utilized in your project because I see no exporting of the preprocessed data. If this is just experimentation, you should indicate that it is and that is if you even need to commit it anyways.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Data-Epic/Best_Nyah/pull/2/discussion/#comment-1373741268'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Bee0933 replied:
actually it was for experimentation.. i'd take it out

Commit:
addressed PR requested changes

- added more doc strings
- refactor function format
- removed redundant file
